Fix #7574: Add support for multi-window & persistent private browsing

Brandon-T opened the pull request:
<!-- *Thank you for submitting a pull request, your contributions are greatly appreciated!* -->

## Security Review
- https://github.com/brave/reviews/issues/1329

## Summary of Changes
- Add support for multi-window
- Add "Open In New Window" and "Open In New Private Window" context menu when long-pressing on a link.
- Add support for persistent private browsing (coming in a separate "commit").
- Added support for Persistent Private Browsing (via a toggle in Privacy Settings), so private tabs can be session restored like Safari and all other browsers.
- Private Browsing ONLY mode will automatically disable persistent private browsing!
- Tabs do NOT save in PBO mode! Turning off PBO will still wipe all your tabs.
- Turning off Persistent Private Browsing (PPB) will not wipe your tabs, but it will wipe them from DISK. So killing the app will not restore them.

## TODO:
- Need a better way to "open a new window" (instead of long-pressing the plus sign, but that's about it, and does not block review).

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #7574

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [ ] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
- Test every feature
- Test private browsing for everything


## Screenshots:

Split (Regular + Regular):
<img width="482" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/e176a473-e349-4629-ba5e-c51b30d46bf7">

Split (Regular + Private):
<img width="490" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/8f0a9026-622b-4c87-bcae-07b56e240721">

Private Window Protection:
<img width="481" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/229a705b-1cc4-48e5-b4d1-502ae98bc376">

Private Window NTP:
<img width="488" alt="image" src="https://github.com/brave/brave-ios/assets/1530031/c63c903b-ed72-453f-892b-e75812915c53">



## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Fix private-tab mode message not showing all the time.
Fix private tab deleting when exiting persistence mode.

kylehickinson reviewed (changes requested):
Runtime Review:

Prefix: These were all tests based on having 2 windows of Brave side-by-side on a 12.9" iPad in landscape orientation

## Bugs That Need Fixing:

- Brave News: Turning on Brave News on one windows NTP breaks Brave News on the other window
  - Expected: News should load on the other window as well and display the same data set

- Brave News: Altering followed sources only brings up "New Content Available" button on window that triggered it. Closing a tab on another window doesn't reload Brave News with the new set of data and will not show the updated data until the app is killed and re-launched.
  - Expected: Both NTPs should reference the same set of News data or at least ensure that new NTP's use the updated data

- Browser Lock: "Unlock" button only appears on 1 window until its tapped on the other?
  - Expected: Should show it on both probably

![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-06-19 at 15 51 07](https://github.com/brave/brave-ios/assets/529104/0d5bbb4c-2644-4109-ac4b-d81247255709)

- Playlist: Plays the same content on multiple windows
  - Expected: If content can only be played on one window at a time then we need to block the other or something with some UI

- Playlist: Shows "Resource Loading" errors when playing from multiple windows

## Polish:

- Shields: Turning shields off for a website loaded in multiple windows does not reflect that
  - Expected: Sync state across windows, reload matching webpages on all connected scenes?

- Brave News: Toggling follow status of a source while viewing the source window on both windows don't update until you leave and revisit the screen
  - Expected: If possible these should be synced so that changes to the 

- Browser Lock: Unlocking from browser lock only unlocks one window. Intended behaviour?
  - Expected: Unlock all scenes that are connected/visible on the screen/foregrounded

- Playlist: Removing an item from Playlist doesn't update the URL bar if you're viewing the videos source webpage at the same time
  - Expected: Should stay in sync so that items added/removed from one window sync to the other

- Tab Tray: For some reason if you open the tab tray on one window and click on the other window, the cells aspect ratio changes.
  - Expected: I assume the longer one is the correct one but not sure

![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-06-19 at 16 09 13](https://github.com/brave/brave-ios/assets/529104/27adbf41-dd57-466e-8afa-8029f84f6a85)

kylehickinson reviewed (changes requested):
This PR may as well been named "Remove PrivateBrowsingManager singleton, oh and by the way also add multi-window support" 😂

Brandon-T committed:
Fix Brave-News not syncing across multiple windows.

Brandon-T committed:
Fix shields not reloading tabs on other windows

Brandon-T committed:
Fix playlist URL bar not updating on another window when added on a different window.

Brandon-T committed:
Adding drag-drop

Brandon-T committed:
Fix private mode search tabs bar disappearing

Brandon-T committed:
Block playlist from showing twice.

Brandon-T committed:
Add persistent private browsing alert for existing users
Move AppState to the same location as delegates
Open context menus like Safari to allow opening URLs in new windows
Moved alert for existing users and updated text.
Updated icons with leo-sfsymbol

Brandon-T committed:
Move private browsing manager to non-singleton.
Fix BraveRewards crash when allocating multiple rewards.
Add persistent private browsing.
Wallet fix for when onboarding is visible in multiple windows.

Fix windowIds being the same on tapping the app icon.
Set window selected
Fix window activation predicate so that tapping the Application Icon does not create new windows.

Added documentation to the SceneDelegate when restoring.

Brandon-T committed:
Add context menu to be on par with Safari.

Brandon-T committed:
Fix unit tests
Rebase & Fixed bug in alert

Brandon-T committed:
Removed preference and popup onboarding the user
Default persistent private browsing to off.

iccub committed:
Rebase fixes.
